server/src/test/java/org/opensearch/cluster/node/DiscoveryNodesTests.java

```diff
@@ -497,6 +511,164 @@ public void testMaxMinNodeVersion() {
         assertEquals(LegacyESVersion.fromString("5.1.0"), build.getMinNodeVersion());
     }
 
+    public void testToXContentInAPIMode() throws IOException {
+        DiscoveryNodes nodes = buildDiscoveryNodes();
+
+        String expectedNodeAPUXContent = "%1$s\"node_%2$d\" : {\n"
+            + "%1$s  \"name\" : \"name_%2$d\",\n"
+            + "%1$s  \"ephemeral_id\" : \"%3$s\",\n"
+            + "%1$s  \"transport_address\" : \"%4$s\",\n"
+            + "%1$s  \"attributes\" : {%5$s}\n"
+            + "%1$s}";
+
+        logger.info(nodes);
+
+        verifyToXContentInContextMode(
+            CONTEXT_MODE_API,
+            nodes,
+            "{\n" + "  \"nodes\" : {\n" + nodes.getNodes().entrySet().stream().map(entry -> {
+                int id = Integer.parseInt(entry.getKey().split("_")[1]);
+                return String.format(
+                    Locale.ROOT,
+                    expectedNodeAPUXContent,
+                    "    ",
+                    id,
+                    entry.getValue().getEphemeralId(),
+                    entry.getValue().getAddress().toString(),
+                    entry.getValue().getAttributes().isEmpty()
+                        ? " "
+                        : "\n" + "        \"custom\" : \"" + entry.getValue().getAttributes().get("custom") + "\"\n      "
+                );
+            }).collect(Collectors.joining(",\n")) + "\n" + "  }\n" + "}"
```

Comment:
Holy readability, Batman!  Switching from concatenation to a stream with some concatenation and `String.format()` combined....  I know it's a test class but even then I'm not sure what it's testing without a comment explaining what this is supposed to look like. This code here might be screaming for a `StringBuilder` implementation. Or a comment like you have below in `getExpectedXContentInGatewayMode`.

Reply:
I know this was bad when I wrote this. To keep the random discovery node, it was really difficult to create a string which was readabile enough. Let me give StringBuilder a try.

Reply:
It's ok as is if you just give an example of what it builds.